File: src/004/main.c
# include <stdbool.h>
# include <stdio.h>

# define MAX 999
# define MIN 100

bool is_palindrome (unsigned int number) {
	unsigned int original = number;
	unsigned int reversed = 0;
	
	while (number) {
		reversed = reversed * 10 + number % 10;
		number /= 10;
	}
	
	return original == reversed;
}

int main () {
	unsigned int largest = 0;
	
	for (int i = MAX; i >= MIN; i--) {
		for (int j = MAX; j >= MIN; j--) {
			unsigned int number = i * j;
			
			if (number < largest) {
				break;
			}
			if (is_palindrome (number)) {
				largest = number;
				break;
			}
		}
	}
	
	printf ("%u\n", largest);
}

File: src/004/main.rs

const MAX: u32 = 999;
const MIN: u32 = 100;

fn is_palindrome (mut number: u32) -> bool {
	let original: u32 = number;
	let mut reversed: u32 = 0;
	
	while number != 0 {
		reversed = reversed * 10 + number % 10;
		number /= 10;
	}
	
	return original == reversed;
}

fn main () {
	let mut largest: u32 = 0;
	
	for i in (MIN..=MAX).rev () {
		for j in (MIN..=MAX).rev () {
			let number: u32 = i * j;
			
			if number < largest {
				break;
			}
			if is_palindrome (number) {
				largest = number;
				break;
			}
		}
	}
	
	println! ("{}", largest);
}

